File: etl/sources/k10_gcd.R
# K10 — Global Charcoal Database: global kompositt-kurve.
#
# Kjøres av etl/sources/k10_gcd.py, som kaller Rscript med utfilen som eneste
# argument. Skriptet gjør én ting: bygger kompositten og skriver den til CSV
# med én gang. Alt annet — enheter, fotnoter, kanonisk form — skjer i Python.
#
# Skriptet skriver CSV-en før det gjør noe som helst annet med resultatet, slik
# at en kompositt som først er beregnet, ikke går tapt om noe senere feiler.
# Pakkene er fra 2019 og 2020, og paleofire ble trukket fra CRAN i 2023.
#
# Metoden er paleofires egen: minimaks-skalering, Box-Cox-transformasjon og
# z-score mot en felles basisperiode, deretter et lavpassfiltrert snitt med
# bootstrappet usikkerhet. Se CLAUDE.md § 5 (K10) og § 7.

args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 1) stop("Bruk: Rscript k10_gcd.R <utfil.csv>")
utfil <- args[1]

suppressPackageStartupMessages({
  library(GCD)
  library(paleofire)
})

cat("K10: GCD", as.character(packageVersion("GCD")),
    "| paleofire", as.character(packageVersion("paleofire")), "\n")

# Alle sedimentære kullserier i databasen. Ingen redaksjonelt utvalg av
# steder — utvalget er «alt som finnes», slik at kurven ikke bærer et valg vi
# ikke kan begrunne maskinelt (CLAUDE.md P1).
alle <- pfSiteSel()
cat("K10: steder i databasen:", length(alle$id_site), "\n")

# Transformasjon til sammenlignbare serier. Basisperioden er den paleofire
# bruker i sine egne eksempler, og dekker en periode med mange serier.
# verbose = TRUE er ikke for loggens skyld. Framdriftssjekken inne i pakken er
# skrevet «k %in% values & verbose == TRUE», og & er ikke kortsluttende: R
# regner ut «k %in% values» uansett. values finnes bare når verbose er TRUE, og
# ellers plukker R opp raster::values — en funksjon — og feiler med at match
# krever vektorargumenter. Slår vi på framdriften, er variabelen definert.
tr <- pfTransform(
  alle,
  method = c("MinMax", "Box-Cox", "Z-Score"),
  BasePeriod = c(200, 4000),
  verbose = TRUE
)

# Lavpassfiltrert kompositt med bootstrappet konfidensintervall. tarAge er
# alderaksen i kalenderår før 1950 (BP): 0 er 1950, negative verdier er årene
# etter 1950.
# hw er halve vindusbredden i år. Den står også i teksten for f_smoothed i
# data/_footnotes.json — endres den her, endres teksten i samme commit (§ 9).
maal_alder <- seq(-60, 8000, by = 20)
komp <- pfCompositeLF(
  tr,
  tarAge = maal_alder,
  binhw = 10,
  hw = 500,
  nboot = 1000,
  verbose = TRUE
)

# Feltnavnene er paleofires egne: BootMean er kurven, BootCi er det
# bootstrappede konfidensintervallet, og antall bidragende serier per bin
# telles av de kolonnene som ikke er NA — samme regnestykke som pakkens egen
# plot(add = "sitenum").
# Formen på det pfCompositeLF gir fra seg, skrives ut før den brukes. En
# kolonne av feil lengde blir resirkulert av data.frame() til en konstant
# kurve, og det ser ut som data i stedet for som en feil.
cat("K10: BinCentres", length(komp$BinCentres),
    "| BootMean", paste(dim(komp$BootMean), collapse="x"),
    "lengde", length(komp$BootMean),
    "| BootCi", paste(dim(komp$BootCi), collapse="x"),
    "| BinnedData", paste(dim(komp$BinnedData), collapse="x"), "\n")
kurve <- as.numeric(komp$BootMean)
cat("K10: kurven har", length(kurve), "punkter,",
    sum(!is.na(kurve)), "uten NA, spenn",
    sprintf("%.4f til %.4f", min(kurve, na.rm=TRUE), max(kurve, na.rm=TRUE)),
    "| unike verdier:", length(unique(round(kurve, 6))), "\n")

ut <- data.frame(
  age_bp = as.numeric(komp$BinCentres),
  composite = kurve,
  lower = as.numeric(komp$BootCi[, 1]),
  upper = as.numeric(komp$BootCi[, 2]),
  n_sites = ncol(komp$BinnedData) - rowSums(is.na(komp$BinnedData))
)

# En kompositt uten variasjon er ikke en kompositt. Stopp før skriving, slik
# at en flat kurve ikke havner i data/processed/ og ser ut som et resultat.
if (length(unique(round(kurve[!is.na(kurve)], 6))) < 2) {
  stop("kompositten er konstant — pfCompositeLF ga ingen variasjon")
}

# Skriv med én gang, før noe annet gjøres med resultatet.
write.csv(ut, utfil, row.names = FALSE)
cat("K10: skrev", nrow(ut), "rader til", utfil, "\n")

File: etl/sources/k10_gcd.py
"""K10 — Global Charcoal Database: kompositt-kurven.

Kjører ``k10_gcd.R``, som bygger kompositten med paleofire og skriver den til
CSV med én gang, og leser CSV-en tilbake. Ingen tolkning, ingen omregning —
kompositten er en z-score og har ingen enhet å regne om (CLAUDE.md § 6).

Kilden er et *proxy*: sedimentært kull er et indirekte spor etter brann, ikke
en måling av brent areal. Serien får alltid ``f_proxy`` og kan aldri tegnes i
samme figur som en km²-serie.

Pakkene
-------
``GCD`` ligger på CRAN. ``paleofire`` ble **trukket fra CRAN 10. januar 2023**
og finnes bare i arkivet. Den siste versjonen, 1.2.4 fra 2019, importerer
``rgdal``, som selv ble trukket i oktober 2023.

``rgdal`` brukes kun av ``pfGridding`` og ``pfToKml``, som vi ikke bruker — det
er ``NAMESPACE`` som krever pakken ved lasting. Installasjonssteget i
workflowen fjerner derfor rgdal fra ``NAMESPACE`` og ``DESCRIPTION`` og de to
filene som bruker den, før pakken installeres. Funksjonene kompositten bygges
av røres ikke. At pakken er lappet, står i ``data/_sources.json``, slik at det
ikke er skjult hvordan tallene ble til.

Kjøres som modul fra repotoppen: ``python -m etl.sources.k10_gcd``
"""

import csv
import hashlib
import json
import subprocess
from datetime import date, datetime, timezone
from pathlib import Path

from etl.schema import RAW_DIR, SOURCES_JSON, STATUS_JSON

SOURCE_ID = "K10"
SERIES_ID = "gcd_charcoal_composite"

R_SKRIPT = Path(__file__).resolve().parent / "k10_gcd.R"
RAW_K10_DIR = RAW_DIR / "k10"
RAW_CSV = RAW_K10_DIR / "gcd_kompositt.csv"

LANDING_URL = "https://cran.r-project.org/package=GCD"
PALEOFIRE_ARKIV_URL = (
    "https://cran.r-project.org/src/contrib/Archive/paleofire/paleofire_1.2.4.tar.gz"
)

# Siteringen paleofire selv oppgir, gjengitt ordrett (CLAUDE.md § 5).
SITERING = (
    "Blarquez, O., Vannière, B., Marlon, J.R., Daniau, A.-L., Power, M.J., "
    "Brewer, S. & Bartlein, P.J. (2014). paleofire: an R package to analyse "
    "sedimentary charcoal records from the Global Charcoal Database to "
    "reconstruct past biomass burning. Computers & Geosciences 72: 255–261."
)

# Kompositten er datert i kalenderår før 1950 (BP), som er standarden i
# paleoklimatologi. Datamodellen fører kalenderår, så alderen regnes om her.
# Det er en datering, ikke en enhet, og hører derfor hjemme i kildemodulen.
BP_NULLPUNKT = 1950


class Kildefeil(Exception):
    """Reises når kompositten ikke lot seg bygge."""


def _sha256(sti):
    sum_ = hashlib.sha256()
    with open(sti, "rb") as f:
        for blokk in iter(lambda: f.read(1 << 20), b""):
            sum_.update(blokk)
    return sum_.hexdigest()


def hent():
    """Kjører R-skriptet og leser kompositten det skrev.

    Returnerer (rader, info). Radene står slik R-skriptet skrev dem, med alder
    i år før 1950.
    """
    RAW_K10_DIR.mkdir(parents=True, exist_ok=True)
    RAW_CSV.unlink(missing_ok=True)

    resultat = subprocess.run(
        ["Rscript", "--vanilla", str(R_SKRIPT), str(RAW_CSV)],
        capture_output=True,
        text=True,
    )
    print(resultat.stdout, end="")
    if resultat.returncode != 0:
        raise Kildefeil(
            "R-skriptet feilet med kode "
            f"{resultat.returncode}. Siste utdata:\n{resultat.stderr[-2000:]}"
        )
    if not RAW_CSV.exists():
        raise Kildefeil(
            f"R-skriptet gikk gjennom, men skrev ingen {RAW_CSV.name}. "
            "Kompositten finnes ikke, og ingenting publiseres."
        )

    with open(RAW_CSV, encoding="utf-8") as f:
        rader = list(csv.DictReader(f))
    if not rader:
        raise Kildefeil(f"{RAW_CSV.name} er tom")

    manglende = {"age_bp", "composite", "n_sites"} - set(rader[0])
    if manglende:
        raise Kildefeil(
            f"{RAW_CSV.name} mangler kolonnene {sorted(manglende)}. "
            f"Den har {sorted(rader[0])}."
        )

    info = {
        "source_id": SOURCE_ID,
        "series_id": SERIES_ID,
        "downloaded_at": date.today().isoformat(),
        "checksum": _sha256(RAW_CSV),
        "rader": len(rader),
    }
    return rader, info


def kalenderaar(age_bp):
    """Gjør alder i år før 1950 om til kalenderår.

    Negative kalenderår er år før vår tidsregning, slik ISO 8601 skriver dem
    med fortegn (CLAUDE.md § 6).
    """
    return BP_NULLPUNKT - int(round(float(age_bp)))


def skriv_metadata(info, processed_files):
    """Registrerer kilden i data/_sources.json."""
    with open(SOURCES_JSON, encoding="utf-8") as f:
        sources = json.load(f)

    sources["sources"][SOURCE_ID] = {
        "source_id": SOURCE_ID,
        "name": "Global Charcoal Database — global kompositt av sedimentært kull",
        "publisher": "Global Paleofire Working Group, via R-pakkene GCD og "
        "paleofire",
        "url": LANDING_URL,
        "download_url": PALEOFIRE_ARKIV_URL,
        "license": "GPL (>= 2) for pakkene. Dataene i GCD er fritt "
        "tilgjengelige med kildeangivelse.",
        "license_url": "https://cran.r-project.org/web/licenses/GPL-2",
        "attribution": SITERING,
        "requires_agreement": False,
        "geography": "global",
        "coverage_start": str(info["aar_forste"]),
        "coverage_end": str(info["aar_siste"]),
        "temporal_resolution": "binned",
        "quality": "reconstructed",
        "unit_source": "zscore",
        "downloaded_at": info["downloaded_at"],
        "source_last_updated": None,
        "checksum": info["checksum"],
        "package_versions": info.get("pakkeversjoner", {}),
        "series": [SERIES_ID],
        "processed_files": processed_files,
        "footnotes": info["footnotes"],
        "notes": "Kompositt bygget med paleofire: minimaks, Box-Cox og z-score "
        "mot en felles basisperiode, deretter lavpassfiltrert snitt med "
        "bootstrappet usikkerhet. Utvalget er alle stedene i databasen, ikke et "
        "redaksjonelt utvalg. paleofire ble trukket fra CRAN i 2023 og "
        "installeres fra arkivet, uten rgdal, som pakken bare bruker i "
        "funksjoner vi ikke kaller. Alderen er regnet om fra år før 1950 til "
        "kalenderår.",
    }
    with open(SOURCES_JSON, "w", encoding="utf-8") as f:
        json.dump(sources, f, ensure_ascii=False, indent=2)
        f.write("\n")


def skriv_status(status, melding, info=None):
    """Skriver kjørestatus til data/_status.json."""
    with open(STATUS_JSON, encoding="utf-8") as f:
        data = json.load(f)

    naa = datetime.now(timezone.utc).isoformat(timespec="seconds")
    forrige = data["sources"].get(SOURCE_ID, {})
    data["last_run"] = naa
    data["sources"][SOURCE_ID] = {
        "status": status,
        "last_attempt": naa,
        "last_success": naa if status == "ok" else forrige.get("last_success"),
        "rows": info["rows"] if info and "rows" in info else forrige.get("rows"),
        "checksum": info["checksum"] if info else forrige.get("checksum"),
        "message": melding,
    }
    with open(STATUS_JSON, "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False, indent=2)
        f.write("\n")


def main():
    rader, info = hent()
    print(f"K10: {info['rader']} rader i kompositten")
    return rader, info


if __name__ == "__main__":
    main()
